File: code/cmt/src/conditionaldistribution.cpp
#include "Eigen/Core"
using Eigen::Dynamic;
using Eigen::Array;
using Eigen::MatrixXd;

#include "conditionaldistribution.h"
#include "exception.h"

#include <cmath>
using std::log;

CMT::ConditionalDistribution::~ConditionalDistribution() {
}



Array<double, 1, Dynamic> CMT::ConditionalDistribution::logLikelihood(
	const pair<ArrayXXd, ArrayXXd>& data) const
{
	return logLikelihood(data.first, data.second);
}



double CMT::ConditionalDistribution::evaluate(
	const MatrixXd& input,
	const MatrixXd& output) const
{
	return -logLikelihood(input, output).mean() / log(2.) / dimOut();
}



double CMT::ConditionalDistribution::evaluate(
	const MatrixXd& input,
	const MatrixXd& output,
	const Preconditioner& preconditioner) const
{
	return -logLikelihood(preconditioner(input, output)).mean() / log(2.) / dimOut()
		- preconditioner.logJacobian(input, output).mean() / log(2.) / dimOut();
}



double CMT::ConditionalDistribution::evaluate(
	const pair<ArrayXXd, ArrayXXd>& data) const
{
	return -logLikelihood(data.first, data.second).mean() / log(2.) / dimOut();
}



double CMT::ConditionalDistribution::evaluate(
	const pair<ArrayXXd, ArrayXXd>& data,
	const Preconditioner& preconditioner) const
{
	return -logLikelihood(preconditioner(data.first, data.second)).mean() / log(2.) / dimOut()
		- preconditioner.logJacobian(data).mean() / log(2.) / dimOut();
}



/**
 * Computes the expectation value of the output.
 */
MatrixXd CMT::ConditionalDistribution::predict(const MatrixXd& input) const {
	throw Exception("This method is not yet implemented.");
}

File: code/cmt/matlab/+cmt/ConditionalDistribution.m
classdef (Abstract) ConditionalDistribution < cmt.MexInterface
    properties (SetAccess = private)
        % Dimensionality of inputs.
        dimIn;
        % Dimensionality of outputs.
        dimOut;
    end

    methods
        function self = ConditionalDistribution(varargin)
            self@cmt.MexInterface(varargin{:});
        end


        function v = get.dimIn(self)
            v = self.mexEval('dimIn');
        end

        function v = get.dimOut(self)
            v = self.mexEval('dimOut');
        end


        function value = sample(self, input)
            %SAMPLE generates outputs for given inputs.
            %   Parameters:
            %       input - inputs stored in columns
            %   Returns:
            %       sampled outputs
            value = self.mexEval('sample', input);
        end

        function value = predict(self, input)
            %PREDICT computes the expectation value of the output.
            %   Parameters:
            %       input - inputs stored in columns
            %   Returns:
            %       expected value of outputs
            value = self.mexEval('predict', input);
        end

        function value = logLikelihood(self, input, output)
            %LOGLIKELIHOOD computes the conditional log-likelihood for the given data points in nats.
            %   Parameters:
            %       input - inputs stored in columns
            %       output - outputs stored in columns
            %   Returns:
            %       log-likelihood of the model evaluated for each data point
            value = self.mexEval('logLikelihood', input, output);
        end

        function loglikelihood = evaluate(self, input, output)
            %EVALUATE computes the average negative conditional log-likelihood for data.
            %   Computed in bits per output component (smaller is better).
            %   If a preconditioner is specified, the data is transformed
            %   before computing the likelihood and the result is corrected
            %   for the Jacobian of the transformation. Note that the data
            %   should *not* already be transformed when specifying a
            %   preconditioner.
            %   Parameters:
            %       input - inputs stored in columns
            %       output - outputs stored in columns
            %       preconditioner - preconditioner that is used to transform the data
            %   Returns:
            %       performance in bits per component
            loglikelihood = self.mexEval('evaluate', input, output);
        end
    end
end
